### PR TITLE
Refactor select_statement and (stacked PR) flatten select after filter/orderby

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Refactors part of select_statements.py to get rid of some flag variables.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 1]
```diff
@@ -557,31 +536,31 @@ def select(self, cols: List[Expression]) -> "SelectStatement":
...
+            new.where = And(self.where, col) if self.where is not None else col
         else:
             new = SelectStatement(
                 from_=self.to_subqueryable(), where=col, analyzer=self.analyzer
             )
+
         return new
 
     def sort(self, cols: List[Expression]) -> "SelectStatement":
+        dependent_columns = derive_dependent_columns(*cols)
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:560]: `dependent_columns` is only used in the `else` branch of line 563. So I prefer leaving it there in the branch.

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 2]
```diff
@@ -557,31 +536,31 @@ def select(self, cols: List[Expression]) -> "SelectStatement":
...
 
     def filter(self, col: Expression) -> "SelectStatement":
-        if self.flatten_disabled:
-            can_be_flattened = False
-        else:
-            dependent_columns = derive_dependent_columns(col)
-            can_be_flattened = can_clause_dependent_columns_flatten(
-                dependent_columns, self.column_states
-            )
+        dependent_columns = derive_dependent_columns(col)
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:539]: Similarly, moving this to the top has extra calculation when `self.flatten_disabled is True`.

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 3]
```diff
@@ -499,46 +499,25 @@ def select(self, cols: List[Expression]) -> "SelectStatement":
...
-                    # query may change sequence of columns. If subquery has same-level reference, flattened sql may not work.
-                    if (
-                        col not in subquery_column_states
-                        or subquery_column_states[col].depend_on_same_level
-                    ):
-                        can_be_flattened = False
-                        break
                     final_projection.append(
-                        copy(subquery_column_states[col].expression)
+                        copy(self.column_states[col].expression)
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:518]: This is a logic change

Comment (sfc-gh-stan) [reply src/snowflake/snowpark/_internal/analyzer/select_statement.py:518]: I deleted line 503 `subquery_column_states = self.column_states` , so these should be equivalent?

File: tests/integ/test_simplifier_suite.py [hunk 4]
```diff
@@ -1028,3 +1029,98 @@ def test_chained_sort(session):
...
+        # Not fully flattened, since col("A") > 1 and col("A") > 2 are referring to different columns
+        (
+            lambda df: df.filter(col("A") > 1)
+            .select((col("B") + 1).alias("A"))
+            .filter(col("A") > 2),
+            'SELECT  *  FROM ( SELECT ("B" + 1 :: INT) AS "A" FROM ( SELECT $1 AS "A", $2 AS "B" FROM  VALUES (1 :: INT, -2 :: INT), (3 :: INT, -4 :: INT)) WHERE ("A" > 1 :: INT)) WHERE ("A" > 2 :: INT)',
+        ),
+        # Not flattened, since we cannot detect dependent columns from sql_expr
+        (
+            lambda df: df.filter(sql_expr("A > 1")).select(col("B"), col("A")),
```
Comment (sfc-gh-yixie) [inline tests/integ/test_simplifier_suite.py:1063]: Better add a test case that uses $.

File: tests/integ/test_simplifier_suite.py [hunk 5]
```diff
@@ -1028,3 +1029,98 @@ def test_chained_sort(session):
...
+        ),
+        # Not flattened, unlike filter, current query takes precendence when there are duplicate column names from a ORDERBY clause
+        (
+            lambda df: df.order_by(col("A")).select((col("B") + 1).alias("A")),
+            'SELECT ("B" + 1 :: INT) AS "A" FROM ( SELECT "A", "B" FROM ( SELECT $1 AS "A", $2 AS "B" FROM  VALUES (1 :: INT, -2 :: INT), (3 :: INT, -4 :: INT)) ORDER BY "A" ASC NULLS FIRST)',
+            True,
+        ),
+        # Not flattened, since we cannot detect dependent columns from sql_expr
+        (
+            lambda df: df.order_by(sql_expr("A")).select(col("B")),
```
Comment (sfc-gh-yixie) [inline tests/integ/test_simplifier_suite.py:1096]: Similarly, add a test case that uses $

**Review by sfc-gh-yixie — APPROVED** [review]

**Comment by sfc-gh-yixie** [comment]
The simplifier logic looks good. But pls check why code coverage is lowered.

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 6]
```diff
@@ -19,6 +19,7 @@
         Analyzer,
     )  # pragma: no cover
 
+
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:22]: Is this generated by precommit hook? 🤔

**Review by sfc-gh-sfan — APPROVED** [review]
It looks pretty neat after refactoring :)

**Comment by sfc-gh-stan** [comment]
> The simplifier logic looks good. But pls check why code coverage is lowered.

Test coverage is lowered since we do not test negating the parameter guard `flatten_select_after_filter_and_orderby`, which should be fine since we plan to remove the temporary parameter anyway.

**Pull request merged**